### DISCUSSION TITLE
Guide: Getting llama.cpp to work on AMD RX 6600 series on Ubuntu via Vulkan for significant performance boost

### DISCUSSION THREAD

**Discussion by liquidscorpio** [ggml-org/llama.cpp#9491] [General]
The guide is about running the Python bindings for llama.cpp - [llama-cpp-python](https://github.com/abetlen/llama-cpp-python) on an RDNA2 series GPU using the Vulkan backend to get ~25x performance boost v/s OpenBLAS on CPU. 

There seems to very sparse information about the topic so writing one here. The steps here should work for vanilla builds of llama.cpp (without the Python bindings) too.

## Context
AMD cards are pain to get working properly in one shot with most LLM frameworks. The RX 6600 family are especially frustrating in this regard as hipBLAS [does not support](https://rocm.docs.amd.com/projects/install-on-windows/en/latest/reference/system-requirements.html) RX 6600 family (LLVM target: gfx1032) of cards. 

While the steps here were tested specifically against RX 6600XT, they should be applicable to other RDNA series cards having Vulkan support.

## Operating System
  
Ubuntu 22.04 LTS (but will work for higher or lower too)

## Install Vulkan

Install Vulkan and OpenBLAS (you may or may not need it).  
More details about APT packages and for other versions can be found [here](https://vulkan.lunarg.com/doc/sdk/1.3.290.0/linux/getting_started_ubuntu.html).
```
wget -qO- https://packages.lunarg.com/lunarg-signing-key-pub.asc | sudo tee /etc/apt/trusted.gpg.d/lunarg.asc
sudo wget -qO /etc/apt/sources.list.d/lunarg-vulkan-jammy.list http://packages.lunarg.com/vulkan/lunarg-vulkan-jammy.list
sudo apt update
sudo apt install vulkan-sdk libopenblas-dev libopenblas0-openmp libopenblas0 libopenblas-dev libopenblas-pthread-dev libopenblas-openmp-dev libopenblas0-openmp libopenblas0-pthread vulkan-tools mesa-utils mesa-utils
```

## Install llama-cpp-python
Install llama-cpp-python with the relevant CMake flags. You can just retain the flags if you are source building llama.cpp and don't need Python. (Works inside a virtualenv too.)

```
CMAKE_ARGS="-DGGML_BLAS=ON -DGGML_BLAS_VENDOR=OpenBLAS -DGGML_VULKAN=1" pip install llama-cpp-python --no-cache-dir --force-reinstall -v
```

Instance the Llama class (Python) and use it
```
llm = Llama(
    model_path=model_path,  # Path to your gguf model file
    verbose=True,  # True is needed for GPU
    n_gpu_layers=-1,  # Set as desired for GPU acceleration
    n_ctx=2048,  # Context window
)
```

## Rough Performance Comparision
| Backend | Hardware | RAM / VRAM | Model | Seconds for ~300 input + output tokens |
| ----- | ----- | ----- | ----- | ----- |
| CPU | Intel i5 8600K | 48G, 2133 Mhz | Mistral 7BQ5 | 80 - 100s |
| GPU | AMD RX 6600XT | 8G DDRR6, 128-bit | Mistral 7BQ5 | 4 - 6s |

**Comment by Stmsmj** [comment]
That was a really good guide. Short and informative. But idk where my problem, i did everything step by step but it threw an error and said cant make the wheel. (Im trying to install it on my anaconda env and i have rx 6600 xt too)

**Comment by liquidscorpio** [comment]
@Stmsmj Possible for you to share the error message?

**Reply by Stmsmj** [reply]
I couldn't install it using pip. I managed to install it using conda-forge but it was an ancient release so it didnt work on my models so i decided to use ollama instead of llama.cpp and surely installation went smoother than llama.cpp😅
Now i can use models with my rx 6600 xt and its hella faster. Made 90 sec to 2 sec.
Anyway tnx for your reply and responsiveness❤

**Reply by liquidscorpio** [reply]
Haha... glad you got the models to work locally using the GPU in the end even if not llama.cpp. Cheers! :)
